crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -328,58 +332,83 @@ fn try_provisioning(
+    let strategy = NodeSetSelectionStrategy::StrictFaultTolerantGreedy;
 
-    let sequencer = preferred_sequencer
+    let preferred_nodes = previous_configuration
+        .map(|p| p.nodeset.clone())
+        .unwrap_or_default();
+
+    let &sequencer = preferred_sequencer
         .and_then(|node_id| {
             // map to a known alive node
             observed_cluster_state.alive_nodes.get(&node_id.id())
         })
         .or_else(|| {
             // we can place the sequencer on any alive node
             observed_cluster_state.alive_nodes.values().choose(&mut rng)
-        });
-
-    if log_servers.len() >= 3 {
-        let replication = ReplicationProperty::new(NonZeroU8::new(2).expect("to be valid"));
-        let mut nodeset = NodeSet::empty();
+        })?;
+
+    let selection = NodeSetSelector::new(nodes_config, observed_cluster_state).select(
+        strategy,
+        &replication,
+        &mut rng,
+        &preferred_nodes,
+    );
+
+    match selection {
+        Ok(selected_nodeset) => {
+            let nodeset = if previous_configuration.map(|p| &p.nodeset).is_some_and(|n| {
+                n.new_excluding(&observed_cluster_state.dead_nodes) == selected_nodeset
+            }) {
+                // The proposed nodeset is the same as the previous one modulo dead nodes; keep it
+                previous_configuration.expect("to exist").nodeset.clone()
+            } else {
+                selected_nodeset
+            };
 
-        for node in &log_servers {
-            nodeset.insert(node.as_plain());
+            Some(ReplicatedLogletParams {
+                loglet_id,
+                sequencer,
+                replication,
+                nodeset,
+                write_set: None,
+            })
         }
 
-        Some(ReplicatedLogletParams {
-            loglet_id,
-            sequencer: *sequencer.expect("one node must exist"),
-            replication,
-            nodeset,
-            write_set: None,
-        })
-    } else if let Some(sequencer) = sequencer {
-        previous_configuration.cloned().map(|mut configuration| {
-            configuration.loglet_id = loglet_id;
-            configuration.sequencer = *sequencer;
-            configuration
-        })
-    } else {
-        None
+        Err(NodeSelectionError::InsufficientWriteableNodes) => {
+            debug!(
+                ?loglet_id,
+                "Insufficient writeable nodes to place replicated loglet"
+            );
+
+            // Although the NodeSetSelector could not find a suitable nodeset, this is also an
+            // opportunity to update the sequencer. This still requires that a quorum of the loglet
+            // nodeset is available in order to seal the previous segment to move to a new sequencer
+            // but make that the data plane's (or, worst case, an operator's) problem.
+            if previous_configuration.is_some_and(|p| p.sequencer != sequencer) {
```

Comment:
The `sequencer` was chosen from the set of alive nodes. So I believe there is no need for this check.

Reply:
Actually with the check for `all_log_servers_alive` in `requires_reconfiguration`, this is potentially dangerous. Imaging three nodes and now a node dies that does not run the sequencer. If we now pick the same sequencer as before, then we will return `None`.

Reply:
I think what then happens is that we try again until we pick the node that was not the sequencer before.